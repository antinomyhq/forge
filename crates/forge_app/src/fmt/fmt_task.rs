use forge_domain::{Status, TaskList};

pub fn to_markdown(before: &TaskList, after: &TaskList) -> String {
    if after.tasks().is_empty() {
        return "No tasks in the list.".to_string();
    }

    let mut markdown: Vec<String> = vec!["\n".to_string()];

    // Create a map of before tasks for comparison
    let before_tasks: std::collections::HashMap<i32, &forge_domain::Task> =
        before.tasks().iter().map(|task| (task.id, task)).collect();

    for task in after.tasks().iter() {
        // Check if task has changed by comparing with before state
        let task_changed = match before_tasks.get(&task.id) {
            Some(before_task) => before_task.status != task.status || before_task.task != task.task,
            None => true, // New task
        };

        let glyph = match task.status {
            Status::Pending => "☐",
            Status::InProgress => "▣",
            Status::Done => "◼",
        };

        let mut text = task.task.clone();

        if task_changed {
            text = format!("**{text}**");
        }

        text = match task.status {
            Status::Pending => text,
            Status::InProgress => format!("__{text}__"),
            Status::Done => format!("~~{text}~~"),
        };

        text = format!("{glyph} {text}");

        markdown.push(text);
    }

    markdown.push("\n".to_owned());

    markdown.join("\n")
}

#[cfg(test)]
mod tests {
    use forge_domain::TaskList;
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn test_to_markdown_empty_task_list() {
        let before = TaskList::new();
        let fixture = TaskList::new();
        let actual = to_markdown(&before, &fixture);
        let expected = "No tasks in the list.";
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_to_markdown_single_pending_task() {
        let before = TaskList::new();
        let mut fixture = TaskList::new();
        fixture.append("Write documentation");
        let actual = to_markdown(&before, &fixture);
        assert!(actual.contains("☐ **Write documentation**")); // New task should be bold
    }

    #[test]
    fn test_to_markdown_mixed_status_tasks() {
        let mut before = TaskList::new();
        let _task1 = before.append("First task");
        let task2 = before.append("Second task");
        before.append("Third task");

        let mut fixture = before.clone();
        // Mark second task as done
        fixture.mark_done(task2.id);
        // Mark first task as in progress manually
        fixture.get_task_mut(0).unwrap().mark_in_progress();

        let actual = to_markdown(&before, &fixture);
        assert!(actual.contains("**First task**")); // changed to in progress
        assert!(actual.contains("**Second task**")); // changed to done
        assert!(actual.contains("☐ Third task")); // unchanged pending
    }

    // Snapshot tests for markdown output
    #[test]
    fn test_empty_task_list_markdown_snapshot() {
        let before = TaskList::new();
        let fixture = TaskList::new();
        let actual = to_markdown(&before, &fixture);
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn test_task_comparison_changes_highlighted() {
        // Test that demonstrates the comparison functionality
        let mut before = TaskList::new();
        let _task1 = before.append("Unchanged task");
        let _task2 = before.append("Task to be modified");
        let task3 = before.append("Task to be completed");

        let mut after = before.clone();
        // Modify task 2's text
        after.get_task_mut(1).unwrap().task = "Modified task text".to_string();
        // Complete task 3
        after.mark_done(task3.id);

        let actual = to_markdown(&before, &after);

        // Unchanged task should not be bold
        assert!(actual.contains("☐ Unchanged task"));
        // Modified task should be bold
        assert!(actual.contains("☐ **Modified task text**"));
        // Completed task should be bold and use done glyph
        assert!(actual.contains("◼ ~~**Task to be completed**~~"));
    }

    #[test]
    fn test_single_pending_task_markdown_snapshot() {
        let before = TaskList::new();
        let mut fixture = TaskList::new();
        fixture.append("Write documentation");
        let actual = to_markdown(&before, &fixture);
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn test_mixed_status_tasks_markdown_snapshot() {
        let mut before = TaskList::new();
        let _task1 = before.append("First task");
        let task2 = before.append("Second task");
        before.append("Third task");

        let mut fixture = before.clone();
        // Mark second task as done
        fixture.mark_done(task2.id);
        // Mark first task as in progress manually
        fixture.get_task_mut(0).unwrap().mark_in_progress();

        let actual = to_markdown(&before, &fixture);
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn test_all_done_tasks_markdown_snapshot() {
        let mut before = TaskList::new();
        let task1 = before.append("Complete feature A");
        let task2 = before.append("Write tests");
        let task3 = before.append("Update documentation");

        let mut fixture = before.clone();
        fixture.mark_done(task1.id);
        fixture.mark_done(task2.id);
        fixture.mark_done(task3.id);

        let actual = to_markdown(&before, &fixture);
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn test_complex_task_list_markdown_snapshot() {
        let mut before = TaskList::new();
        let _task1 = before.append("Review pull request #123");
        let _task2 = before.append("Fix bug in authentication");
        let task3 = before.append("Deploy to staging");
        let _task4 = before.append("Update API documentation");
        let _task5 = before.append("Refactor user service");

        let mut fixture = before.clone();
        // Mark one task as done
        fixture.mark_done(task3.id);
        // Mark two tasks as in progress manually (without removing them)
        fixture.get_task_mut(0).unwrap().mark_in_progress(); // "Review pull request #123"
        fixture.get_task_mut(4).unwrap().mark_in_progress(); // "Refactor user service"

        let actual = to_markdown(&before, &fixture);
        insta::assert_snapshot!(actual);
    }
}
