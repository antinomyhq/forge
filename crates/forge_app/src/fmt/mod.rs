pub mod content;
pub mod fmt_input;
pub mod fmt_output;
pub mod fmt_task;
