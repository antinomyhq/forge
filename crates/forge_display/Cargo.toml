[package]
name = "forge_display"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true
colored.workspace = true
chrono.workspace = true
similar.workspace = true
console.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
pretty_assertions.workspace = true
strip-ansi-escapes.workspace = true