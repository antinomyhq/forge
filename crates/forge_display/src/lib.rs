pub mod diff;
pub mod grep;
pub mod title;

pub use diff::DiffFormat;
pub use grep::GrepFormat;
pub use title::TitleFormat;
