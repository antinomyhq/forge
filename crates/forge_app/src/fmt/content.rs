use forge_display::TitleFormat;
use forge_domain::{ChatResponse, Environment};

#[derive(Debug, PartialEq)]
pub enum ContentFormat {
    Title(TitleFormat),
    PlainText(String),
    Markdown(String),
}

impl From<ContentFormat> for ChatResponse {
    fn from(value: ContentFormat) -> Self {
        match value {
            ContentFormat::Title(title) => {
                ChatResponse::Text { text: title.to_string(), is_complete: true, is_md: false }
            }
            ContentFormat::PlainText(text) => {
                ChatResponse::Text { text, is_complete: true, is_md: false }
            }
            ContentFormat::Markdown(text) => {
                ChatResponse::Text { text, is_complete: true, is_md: true }
            }
        }
    }
}

impl From<TitleFormat> for ContentFormat {
    fn from(title: TitleFormat) -> Self {
        ContentFormat::Title(title)
    }
}

pub trait FormatContent {
    fn to_content(&self, env: &Environment) -> Option<ContentFormat>;
}
