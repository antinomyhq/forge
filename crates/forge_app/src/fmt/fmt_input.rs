use std::convert::AsRef;
use std::path::Path;

use forge_display::TitleFormat;
use forge_domain::{Environment, Tools};

use crate::fmt::content::{ContentFormat, FormatContent};
use crate::utils::display_path;

impl FormatContent for Tools {
    fn to_content(&self, env: &Environment) -> Option<ContentFormat> {
        let display_path_for = |path: &str| display_path(env, Path::new(path));

        let output = match self {
            Tools::ForgeToolFsRead(input) => {
                let display_path = display_path_for(&input.path);
                let is_explicit_range = input.start_line.is_some() || input.end_line.is_some();
                let mut subtitle = display_path;
                if is_explicit_range {
                    match (&input.start_line, &input.end_line) {
                        (Some(start), Some(end)) => {
                            subtitle.push_str(&format!(" [Range {start}-{end}]"));
                        }
                        (Some(start), None) => {
                            subtitle.push_str(&format!(" [Range {start}-]"));
                        }
                        (None, Some(end)) => {
                            subtitle.push_str(&format!(" [Range -{end}]"));
                        }
                        (None, None) => {}
                    }
                };
                TitleFormat::debug("Read").sub_title(subtitle).into()
            }
            Tools::ForgeToolFsCreate(input) => {
                let display_path = display_path_for(&input.path);
                let title = if input.overwrite {
                    "Overwrite"
                } else {
                    "Create"
                };
                TitleFormat::debug(title).sub_title(display_path).into()
            }
            Tools::ForgeToolFsSearch(input) => {
                let formatted_dir = display_path_for(&input.path);
                let title = match (&input.regex, &input.file_pattern) {
                    (Some(regex), Some(pattern)) => {
                        format!("Search for '{regex}' in '{pattern}' files at {formatted_dir}")
                    }
                    (Some(regex), None) => format!("Search for '{regex}' at {formatted_dir}"),
                    (None, Some(pattern)) => format!("Search for '{pattern}' at {formatted_dir}"),
                    (None, None) => format!("Search at {formatted_dir}"),
                };
                TitleFormat::debug(title).into()
            }
            Tools::ForgeToolFsRemove(input) => {
                let display_path = display_path_for(&input.path);
                TitleFormat::debug("Remove").sub_title(display_path).into()
            }
            Tools::ForgeToolFsPatch(input) => {
                let display_path = display_path_for(&input.path);
                TitleFormat::debug(input.operation.as_ref())
                    .sub_title(display_path)
                    .into()
            }
            Tools::ForgeToolFsUndo(input) => {
                let display_path = display_path_for(&input.path);
                TitleFormat::debug("Undo").sub_title(display_path).into()
            }
            Tools::ForgeToolProcessShell(input) => {
                TitleFormat::debug(format!("Execute [{}]", env.shell))
                    .sub_title(&input.command)
                    .into()
            }
            Tools::ForgeToolNetFetch(input) => {
                TitleFormat::debug("GET").sub_title(&input.url).into()
            }
            Tools::ForgeToolFollowup(input) => TitleFormat::debug("Follow-up")
                .sub_title(&input.question)
                .into(),
            Tools::ForgeToolAttemptCompletion(input) => {
                ContentFormat::Markdown(input.result.clone())
            }
            Tools::ForgeToolTaskListAppend(_) => {
                TitleFormat::debug("Task +1 ADD".to_string()).into()
            }
            Tools::ForgeToolTaskListAppendMultiple(input) => {
                TitleFormat::debug(format!("Task +{} ADD", input.tasks.len())).into()
            }
            Tools::ForgeToolTaskListUpdate(_) => {
                TitleFormat::debug("Task Update".to_string()).into()
            }
            Tools::ForgeToolTaskListList(_) => TitleFormat::debug("Task Read".to_string()).into(),
            Tools::ForgeToolTaskListClear(_) => TitleFormat::debug("Task Clear".to_string()).into(),
        };

        Some(output)
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use console::strip_ansi_codes;
    use forge_domain::{Environment, FSRead, FSWrite, Shell, Tools};
    use pretty_assertions::assert_eq;
    use url::Url;

    use super::{ContentFormat, FormatContent};

    impl ContentFormat {
        pub fn render(&self, with_timestamp: bool) -> String {
            match self {
                ContentFormat::Title(title) => title.render(with_timestamp),
                ContentFormat::PlainText(summary) => summary.clone(),
                ContentFormat::Markdown(summary) => summary.clone(),
            }
        }
    }

    fn fixture_environment() -> Environment {
        Environment {
            os: "linux".to_string(),
            pid: 12345,
            cwd: PathBuf::from("/home/user/project"),
            home: Some(PathBuf::from("/home/user")),
            shell: "/bin/bash".to_string(),
            base_path: PathBuf::from("/home/user/project"),
            retry_config: forge_domain::RetryConfig {
                initial_backoff_ms: 1000,
                min_delay_ms: 500,
                backoff_factor: 2,
                max_retry_attempts: 3,
                retry_status_codes: vec![429, 500, 502, 503, 504],
                max_delay: None,
            },
            max_search_lines: 25,
            fetch_truncation_limit: 55,
            max_read_size: 10,
            stdout_max_prefix_length: 10,
            stdout_max_suffix_length: 10,
            http: Default::default(),
            max_file_size: 0,
            forge_api_url: Url::parse("http://forgecode.dev/api").unwrap(),
        }
    }

    #[test]
    fn test_fs_read_basic() {
        let fixture = Tools::ForgeToolFsRead(FSRead {
            path: "/home/user/project/src/main.rs".to_string(),
            start_line: None,
            end_line: None,
            explanation: None,
        });
        let env = fixture_environment();

        let actual_content = fixture.to_content(&env);
        let rendered = actual_content.unwrap().render(false);
        let actual = strip_ansi_codes(&rendered);
        let expected = "⏺ Read src/main.rs";

        assert_eq!(actual, expected);
    }

    #[test]
    fn test_fs_read_with_range() {
        let fixture = Tools::ForgeToolFsRead(FSRead {
            path: "/home/user/project/src/main.rs".to_string(),
            start_line: Some(10),
            end_line: Some(20),
            explanation: None,
        });
        let env = fixture_environment();

        let actual_content = fixture.to_content(&env);
        let rendered = actual_content.unwrap().render(false);
        let actual = strip_ansi_codes(&rendered);
        let expected = "⏺ Read src/main.rs [Range 10-20]";

        assert_eq!(actual, expected);
    }

    #[test]
    fn test_fs_create_new_file() {
        let fixture = Tools::ForgeToolFsCreate(FSWrite {
            path: "/home/user/project/new_file.txt".to_string(),
            content: "Hello world".to_string(),
            overwrite: false,
            explanation: None,
        });
        let env = fixture_environment();

        let actual_content = fixture.to_content(&env);
        let rendered = actual_content.unwrap().render(false);
        let actual = strip_ansi_codes(&rendered);
        let expected = "⏺ Create new_file.txt";

        assert_eq!(actual, expected);
    }

    #[test]
    fn test_fs_create_overwrite() {
        let fixture = Tools::ForgeToolFsCreate(FSWrite {
            path: "/home/user/project/existing_file.txt".to_string(),
            content: "Updated content".to_string(),
            overwrite: true,
            explanation: None,
        });
        let env = fixture_environment();

        let actual_content = fixture.to_content(&env);
        let rendered = actual_content.unwrap().render(false);
        let actual = strip_ansi_codes(&rendered);
        let expected = "⏺ Overwrite existing_file.txt";

        assert_eq!(actual, expected);
    }

    #[test]
    fn test_shell_command() {
        let fixture = Tools::ForgeToolProcessShell(Shell {
            command: "ls -la".to_string(),
            cwd: PathBuf::from("/home/user/project"),
            keep_ansi: false,
            explanation: None,
        });
        let env = fixture_environment();

        let actual_content = fixture.to_content(&env);
        let rendered = actual_content.unwrap().render(false);
        let actual = strip_ansi_codes(&rendered);
        let expected = "⏺ Execute [/bin/bash] ls -la";

        assert_eq!(actual, expected);
    }

    #[test]
    fn test_render_with_and_without_timestamp() {
        let fixture = Tools::ForgeToolFsRead(FSRead {
            path: "/home/user/project/src/main.rs".to_string(),
            start_line: None,
            end_line: None,
            explanation: None,
        });
        let env = fixture_environment();
        let content = fixture.to_content(&env).unwrap();

        // Test render(false) - should not include timestamp
        let rendered_without = content.render(false);
        let actual_without = strip_ansi_codes(&rendered_without);
        assert!(!actual_without.contains("["));
        assert!(!actual_without.contains(":"));

        // Test render(true) - should include timestamp
        let rendered_with = content.render(true);
        let actual_with = strip_ansi_codes(&rendered_with);
        assert!(actual_with.contains("["));
        assert!(actual_with.contains(":"));
    }
}
