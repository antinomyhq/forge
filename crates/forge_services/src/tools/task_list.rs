use std::fmt::Display;
use std::sync::Arc;

use anyhow::Result;
use derive_setters::Setters;
use forge_display::TitleFormat;
use forge_domain::{
    EnvironmentService, ExecutableTool, NamedTool, TaskListInput, ToolCallContext, ToolDescription,
    ToolName,
};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

use crate::{FsWriteService, Infrastructure};

/// Represents the status of a task in the TaskList.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
pub enum TaskStatus {
    /// Task is waiting to be started.
    Pending,
    /// Task is currently being worked on.
    InProgress,
    /// Task has been completed.
    Done,
}

impl Display for TaskStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TaskStatus::Done => write!(f, "Done"),
            TaskStatus::InProgress => write!(f, "InProgress"),
            TaskStatus::Pending => write!(f, "Pending"),
        }
    }
}

/// Represents a single task in the TaskList.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Setters)]
#[setters(into, strip_option)]
pub struct Task {
    /// Unique identifier for the task.
    pub id: u32,
    /// Description of the task.
    pub task: String,
    /// Current status of the task.
    pub status: TaskStatus,
}

impl Display for Task {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "<task id=\"{}\">\n<content>{}</content>\n<status>{}</status>\n</task>",
            self.id, self.task, self.status
        )
    }
}

/// Statistics about the TaskList.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
pub struct Stats {
    /// Total number of tasks in the list.
    pub total_tasks: u32,
    /// Number of completed tasks.
    pub done_tasks: u32,
    /// Number of pending tasks.
    pub pending_tasks: u32,
    /// Number of in-progress tasks.
    pub in_progress_tasks: u32,
}

impl std::fmt::Display for Stats {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "<stats>\n<total_tasks>{}</total_tasks>\n<done_tasks>{}</done_tasks>\n<pending_tasks>{}</pending_tasks>\n<in_progress_tasks>{}</in_progress_tasks>\n</stats>",
            self.total_tasks, self.done_tasks, self.pending_tasks, self.in_progress_tasks)
    }
}

/// TaskList operation result.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct TaskListResult {
    /// Message describing the result of the operation.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    /// The task that was affected by the operation, if applicable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_task: Option<Task>,
    /// Statistics about the task list.
    pub stats: Stats,
    /// List of tasks
    pub tasks: Option<Vec<Task>>,
}

impl Display for TaskListResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut result = String::from("<task_list_result>\n");

        if let Some(message) = &self.message {
            result.push_str(&format!("<message>{message}</message>\n"));
        }

        if let Some(task) = &self.next_task {
            result.push_str("<next_task>\n");
            result.push_str(&format!("{task}\n"));
            result.push_str("\n</next_task>\n");
        }

        result.push_str(&format!("{}\n", self.stats));

        if let Some(tasks) = &self.tasks {
            if !tasks.is_empty() {
                result.push_str("<tasks_list>\n");
                for task in tasks {
                    result.push_str(&format!("{task}\n"));
                }
                result.push_str("</tasks_list>\n");
            }
        }

        result.push_str("</task_list_result>");

        write!(f, "{result}")
    }
}

/// A stateful task management tool that maintains an ordered list of tasks with
/// status tracking. Provides operations to add tasks (append/prepend), mark
/// tasks as in-progress (pop_front/pop_back), complete tasks (mark_done), and
/// view the current state (list). Automatically identifies the next pending
/// task when completing items and provides detailed statistics on task status.
/// Ideal for sequential workflows, project planning, and tracking multi-step
/// processes.
#[derive(Debug, ToolDescription)]
pub struct TaskList<F> {
    infra: Arc<F>,
    tasks: Arc<Mutex<Vec<Task>>>,
    next_id: Arc<Mutex<u32>>,
}

impl<F: Infrastructure> TaskList<F> {
    /// Creates a new TaskList tool with the given infrastructure.
    pub fn new(infra: Arc<F>) -> Self {
        Self {
            infra,
            tasks: Arc::new(Mutex::new(Vec::new())),
            next_id: Arc::new(Mutex::new(1)),
        }
    }

    /// Format the tasks as a markdown list with checkboxes
    pub async fn format_tasks_markdown(&self) -> String {
        let tasks = self.tasks.lock().await;

        if tasks.is_empty() {
            return "No tasks available.".to_string();
        }

        let mut markdown = String::new();

        // Sort tasks by ID for consistent display
        let mut sorted_tasks = tasks.clone();
        sorted_tasks.sort_by_key(|task| task.id);

        for task in sorted_tasks {
            let checkbox = match task.status {
                TaskStatus::Done => "[x]",
                TaskStatus::Pending => "[ ]",
                TaskStatus::InProgress => "[ ]",
            };

            let formatted_task = match task.status {
                TaskStatus::Done => format!("- {} {}", checkbox, task.task),
                TaskStatus::Pending => format!("- {} {}", checkbox, task.task),
                TaskStatus::InProgress => {
                    format!("- {} __{} (In Progress)__ ", checkbox, task.task)
                }
            };

            markdown.push_str(&formatted_task);
            markdown.push('\n');
        }

        markdown
    }

    /// Calculate statistics for the current task list.
    async fn calculate_stats(&self) -> Stats {
        let tasks = self.tasks.lock().await;
        let total_tasks = tasks.len() as u32;
        let done_tasks = tasks
            .iter()
            .filter(|task| task.status == TaskStatus::Done)
            .count() as u32;
        let in_progress_tasks = tasks
            .iter()
            .filter(|task| task.status == TaskStatus::InProgress)
            .count() as u32;
        let pending_tasks = tasks
            .iter()
            .filter(|task| task.status == TaskStatus::Pending)
            .count() as u32;

        Stats { total_tasks, done_tasks, pending_tasks, in_progress_tasks }
    }

    /// Get the next available task ID and increment the counter.
    async fn get_next_id(&self) -> u32 {
        let mut next_id = self.next_id.lock().await;
        let id = *next_id;
        *next_id += 1;
        id
    }

    /// Find the next pending task in the list.
    async fn find_next_pending_task(&self) -> Option<Task> {
        let tasks = self.tasks.lock().await;
        tasks
            .iter()
            .find(|task| task.status == TaskStatus::Pending)
            .cloned()
    }

    /// Handle the append operation, adding a task to the end of the list.
    async fn append(&self, task_desc: String) -> Result<TaskListResult> {
        let task = Task {
            id: self.get_next_id().await,
            task: task_desc,
            status: TaskStatus::Pending,
        };

        {
            let mut tasks = self.tasks.lock().await;
            tasks.push(task.clone());
        }

        let stats = self.calculate_stats().await;

        Ok(TaskListResult {
            message: Some("Task added to the end of the list.".to_string()),
            next_task: None,
            stats,
            tasks: None,
        })
    }

    /// Handle the prepend operation, adding a task to the beginning of the
    /// list.
    async fn prepend(&self, task_desc: String) -> Result<TaskListResult> {
        let task = Task {
            id: self.get_next_id().await,
            task: task_desc,
            status: TaskStatus::Pending,
        };

        {
            let mut tasks = self.tasks.lock().await;
            tasks.insert(0, task.clone());
        }

        let stats = self.calculate_stats().await;

        Ok(TaskListResult {
            message: Some("Task added to the beginning of the list.".to_string()),
            next_task: None,
            stats,
            tasks: None,
        })
    }

    /// Handle the pop_front operation, marking the first pending task as
    /// InProgress without removing it.
    async fn pop_front(&self) -> Result<TaskListResult> {
        let task_option = {
            let mut tasks = self.tasks.lock().await;
            if tasks.is_empty() {
                None
            } else {
                let pending_index = tasks
                    .iter()
                    .position(|task| task.status == TaskStatus::Pending);

                if let Some(index) = pending_index {
                    tasks[index].status = TaskStatus::InProgress;
                    Some(tasks[index].clone())
                } else {
                    None
                }
            }
        };

        let stats = self.calculate_stats().await;

        if let Some(task) = task_option {
            Ok(TaskListResult { message: None, next_task: Some(task), stats, tasks: None })
        } else {
            Ok(TaskListResult {
                message: Some("No pending tasks found.".to_string()),
                next_task: None,
                stats: self.calculate_stats().await,
                tasks: None,
            })
        }
    }

    /// Handle the pop_back operation, marking the last pending task as
    /// InProgress without removing it.
    async fn pop_back(&self) -> Result<TaskListResult> {
        let task_option = {
            let mut tasks = self.tasks.lock().await;
            if tasks.is_empty() {
                None
            } else {
                let pending_task_index = tasks
                    .iter()
                    .rposition(|task| task.status == TaskStatus::Pending);

                if let Some(index) = pending_task_index {
                    tasks[index].status = TaskStatus::InProgress;
                    Some(tasks[index].clone())
                } else {
                    None
                }
            }
        };

        let stats = self.calculate_stats().await;

        if let Some(task) = task_option {
            Ok(TaskListResult { message: None, next_task: Some(task), stats, tasks: None })
        } else {
            Ok(TaskListResult {
                message: Some("No pending tasks found.".to_string()),
                next_task: None,
                stats,
                tasks: None,
            })
        }
    }

    /// Handle the mark_done operation, marking a task as completed and finding
    /// the next pending task.
    async fn mark_done(&self, id: u32) -> Result<TaskListResult> {
        let found = {
            let mut tasks = self.tasks.lock().await;
            let mut found = false;

            for task in tasks.iter_mut() {
                if task.id == id {
                    task.status = TaskStatus::Done;
                    found = true;
                    break;
                }
            }

            found
        };

        if !found {
            return Ok(TaskListResult {
                message: Some(format!("No task found with ID {id}.")),
                next_task: None,
                stats: self.calculate_stats().await,
                tasks: None,
            });
        }

        let next_task = self.find_next_pending_task().await;
        let stats = self.calculate_stats().await;

        Ok(TaskListResult {
            message: if next_task.is_some() {
                Some(format!(
                    "Task {} marked as done. Next taskId is {}.",
                    id,
                    next_task.as_ref().unwrap().id
                ))
            } else {
                Some(format!("Task {id} marked as done. No more pending tasks."))
            },
            next_task,
            stats,
            tasks: None,
        })
    }

    /// Handle the list operation, returning all tasks in the list.
    async fn list(&self) -> Result<TaskListResult> {
        let stats = self.calculate_stats().await;
        let tasks = self.tasks.lock().await;
        Ok(TaskListResult {
            message: None,
            next_task: None,
            stats,
            tasks: Some(tasks.clone()),
        })
    }
}

impl<F> NamedTool for TaskList<F> {
    fn tool_name() -> ToolName {
        ToolName::new("forge_tool_task_list")
    }
}

fn format_input(input: &TaskListInput) -> String {
    if let Some(task) = &input.append_task {
        return format!("Append: {task}");
    } else if let Some(task) = &input.prepend_task {
        return format!("Prepend: {task}");
    } else if input.pop_front.is_some() {
        return "PopFront".to_string();
    } else if input.pop_back.is_some() {
        return "PopBack".to_string();
    } else if let Some(id) = input.mark_done_id {
        return format!("MarkDone: {id}");
    } else if input.list.is_some() {
        return "List".to_string();
    }

    "Unknown operation".to_string()
}

#[async_trait::async_trait]
impl<F: Infrastructure> ExecutableTool for TaskList<F> {
    type Input = TaskListInput;

    async fn call(&self, mut context: ToolCallContext, input: Self::Input) -> Result<String> {
        context
            .send_text(TitleFormat::debug("TaskList").sub_title(format_input(&input)))
            .await?;

        let result = if let Some(task) = input.append_task {
            self.append(task).await?
        } else if let Some(task) = input.prepend_task {
            self.prepend(task).await?
        } else if input.pop_front.is_some() {
            self.pop_front().await?
        } else if input.pop_back.is_some() {
            self.pop_back().await?
        } else if let Some(id) = input.mark_done_id {
            self.mark_done(id).await?
        } else if input.list.is_some() {
            self.list().await?
        } else {
            return Err(anyhow::anyhow!("No valid operation specified"));
        };

        // Format tasks in markdown format
        let markdown = self.format_tasks_markdown().await;
        let cwd = self
            .infra
            .environment_service()
            .get_environment()
            .cwd
            .join("task_list.md");
        self.infra
            .file_write_service()
            .write(cwd.as_path(), markdown.into())
            .await?;

        context
            .set_stats(self.calculate_stats().await.to_string())
            .await;

        Ok(result.to_string())
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use forge_domain::ToolCallContext;

    use super::*;
    use crate::attachment::tests::MockInfrastructure;

    fn create_task_list() -> TaskList<MockInfrastructure> {
        let infra = Arc::new(MockInfrastructure::new());
        TaskList::new(infra)
    }

    #[tokio::test]
    async fn test_append_task() {
        // Create fixture
        let task_list = create_task_list();
        let input = TaskListInput {
            append_task: Some("Test task".to_string()),
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: None,
        };

        // Execute the fixture
        let result = task_list
            .call(ToolCallContext::default(), input)
            .await
            .unwrap();

        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_prepend_task() {
        // Create fixture
        let task_list = create_task_list();

        // Add two tasks
        let input1 = TaskListInput {
            append_task: Some("Task 1".to_string()),
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: None,
        };
        let input2 = TaskListInput {
            append_task: None,
            prepend_task: Some("Task 2".to_string()),
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: None,
        };

        // Execute the fixture
        task_list
            .call(ToolCallContext::default(), input1)
            .await
            .unwrap();
        let result = task_list
            .call(ToolCallContext::default(), input2)
            .await
            .unwrap();

        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_pop_front() {
        // Create fixture
        let task_list = create_task_list();

        // Add a task and mark it as in-progress
        let input1 = TaskListInput {
            append_task: Some("Task to pop".to_string()),
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: None,
        };
        let input2 = TaskListInput {
            append_task: None,
            prepend_task: None,
            pop_front: Some(true),
            pop_back: None,
            mark_done_id: None,
            list: None,
        };

        // Execute the fixture
        task_list
            .call(ToolCallContext::default(), input1)
            .await
            .unwrap();
        let result = task_list
            .call(ToolCallContext::default(), input2)
            .await
            .unwrap();

        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_pop_back() {
        // Create fixture
        let task_list = create_task_list();

        // Add two tasks and mark the last one as in-progress
        let input1 = TaskListInput {
            append_task: Some("Task 1".to_string()),
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: None,
        };
        let input2 = TaskListInput {
            append_task: Some("Task 2".to_string()),
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: None,
        };
        let input3 = TaskListInput {
            append_task: None,
            prepend_task: None,
            pop_front: None,
            pop_back: Some(true),
            mark_done_id: None,
            list: None,
        };

        // Execute the fixture
        task_list
            .call(ToolCallContext::default(), input1)
            .await
            .unwrap();
        task_list
            .call(ToolCallContext::default(), input2)
            .await
            .unwrap();
        let result = task_list
            .call(ToolCallContext::default(), input3)
            .await
            .unwrap();

        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_mark_done() {
        // Create fixture
        let task_list = create_task_list();

        // Add two tasks, mark first as done
        let input1 = TaskListInput {
            append_task: Some("Task 1".to_string()),
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: None,
        };

        // Execute the fixture
        task_list
            .call(ToolCallContext::default(), input1)
            .await
            .unwrap();

        // For testing purposes, we'll use ID 1 for the first task
        let id = 1;
        let input3 = TaskListInput {
            append_task: None,
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: Some(id),
            list: None,
        };
        let result = task_list
            .call(ToolCallContext::default(), input3)
            .await
            .unwrap();

        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    async fn test_empty_list() {
        // Create fixture
        let task_list = create_task_list();
        let input = TaskListInput {
            append_task: None,
            prepend_task: None,
            pop_front: None,
            pop_back: None,
            mark_done_id: None,
            list: Some(true),
        };

        // Execute the fixture
        let actual = task_list
            .call(ToolCallContext::default(), input)
            .await
            .unwrap();

        insta::assert_snapshot!(actual);
    }
}
