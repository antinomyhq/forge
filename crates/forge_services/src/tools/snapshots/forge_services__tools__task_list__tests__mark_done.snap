---
source: crates/forge_services/src/tools/task_list.rs
expression: result
---
<task_list_result>
<message>Task 0 marked as done. No more pending tasks.</message>
<stats>
<total_tasks>1</total_tasks>
<done_tasks>1</done_tasks>
<pending_tasks>0</pending_tasks>
<in_progress_tasks>0</in_progress_tasks>
</stats>
</task_list_result>
